sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -83,7 +83,8 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
   }
 
   {
-    std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
+    std::unique_lock<std::shared_timed_mutex> Lock(MGraphLock, std::defer_lock);
+    lockSharedTimedMutex(Lock);
 
     // TODO: Check if lazy mode.
```

Comment:
You need to check that NewCmd on line 91 is still valid as it can be executed and removed between locks.
Suggest storing an event instead of command on line 66, then extract and check that it is non-null the command from the event on line 90.

Also line 97 looks not thread safe.

Reply:
Ok, I will do that.

Reply:
As we decided to use a single lock in the function Scheduler::addCG(), we no longer need to check NewCmd on line 91.

I checked line 97,  flushStreams() calls Scheduler::addHostAccessor() inside itself that locks MGraphLock.